tools/run_tests/xds_k8s_test_driver/framework/test_app/client_app.py

```diff
@@ -419,23 +435,65 @@ def find_subchannels_with_state(
-    ) -> _ChannelzChannel:
-        """Checks if the channel has any successful calls.
-
-        We consider the channel is active if channel is in READY state and calls_started is
-        greater than calls_failed.
+    def check_channel_in_flight_calls(
+        self,
+        channel: _ChannelzChannel,
+        *,
+        wait_between_checks: Optional[_timedelta] = None,
+        **rpc_params,
+    ) -> Optional[_ChannelzChannel]:
+        """Checks if the channel has calls that started, but didn't complete.
+
+        We consider the channel is active if channel is in READY state and
+        calls_started is greater than calls_failed.
+
+        This method address race where a call to the xDS control plane server
+        has just started and a channelz request comes in before the call has
+        had a chance to fail.
+
+        With channels to the xDS control plane, the channel can be READY but the
+        calls could be failing to initialize, f.e. due to a failure to fetch
+        OAUTH2 token. To increase the confidence that we have a valid channel
+        with working OAUTH2 tokens, we check whether the channel is in a READY
+        state with active calls twice with an interval of 2 seconds between the
+        two attempts. If the OAUTH2 token is not valid, the call would fail and
+        be caught in either the first attempt, or the second attempt. It is
+        possible that between the two attempts, a call fails and a new call is
+        started, so we also test for equality between the started calls of the
+        two channelz results.
+
+        There still exists a possibility that a call fails on fetching OAUTH2
+        token after 2 seconds (maybe because there is a slowdown in the
+        system.) If such a case is observed, consider increasing the interval
+        from 2 seconds to 5 seconds.
+
+        Returns updated channel on success, or None on failure.
         """
+        if not self.calc_calls_in_flight(channel):
+            return None
+
+        if not wait_between_checks:
+            wait_between_checks = _timedelta(seconds=2)
+
+        # Load the channel second time after the timeout.
+        time.sleep(wait_between_checks.total_seconds())
+        channel_upd: _ChannelzChannel = self.channelz.get_channel(
+            channel.ref.channel_id, **rpc_params
+        )
         if (
-            channel.data.state.state is _ChannelzChannelState.READY
-            and channel.data.calls_started > channel.data.calls_failed
+            not self.calc_calls_in_flight(channel_upd)
+            or channel.data.calls_started != channel_upd.data.calls_started
         ):
-            return channel
+            return None
+        return channel_upd
 
-        raise self.NotFound(
-            f"[{self.hostname}] Not found successful calls over the channel."
-        )
+    @classmethod
+    def calc_calls_in_flight(cls, channel: _ChannelzChannel) -> int:
```

Comment:
If we're just checking whether return is 0, we should consider change this to return `bool`. Otherwise can we log the number somewhere?

Reply:
The logic is pretty much the same, except with converting to bool at the end. I think this method could be useful as is. F.e. I can use it in ipython:

```pycon
In [1]: from framework.test_app import client_app
In [2]: client = client_app.XdsTestClient(ip='127.0.0.1', rpc_port=8079, server_target='xds:///xds-test-server', hostname='sergiitk-client-dev')
In [3]: xds_chan = client.find_active_xds_channel(client_app.DEFAULT_TD_XDS_URI)
In [4]: client.calc_calls_in_flight(xds_chan)
Out[4]: 2
```

Reply:
I see, make sense.